content/post/ddos-detection.md

```diff
@@ -0,0 +1,419 @@
+
+```python
+import json
+import pickle
+import struct
+
+import wallaroo
+
+from circular_buffer import CircularBuffer
+from stats import weighted_mu_sigma
+
+
+def application_setup(args):
+    in_host, in_port = wallaroo.tcp_parse_input_addrs(args)[0]
+    out_host, out_port = wallaroo.tcp_parse_output_addrs(args)[0]
+
+    partitions = 10000
+    server_partitions = range(partitions)
+
+    ab = wallaroo.ApplicationBuilder("DDoS Attack Detector")
+    ab.new_pipeline("ddos detection",
+                    wallaroo.TCPSourceConfig(in_host, in_port, Decoder()))
+    ab.to_state_partition_u64(ProcessLogEntry(), PartitionStatsBuilder(),
+                          "process log entry",
+                          ServerPartitionFunction(partitions),
+                          server_partitions)
+    ab.to_sink(wallaroo.TCPSinkConfig(out_host, out_port, Encoder()))
+    return ab.build()
+```
+
+## Running the Application
+
+To run the application, we will use netcat for our TCP receiver, and a custom sender that's included in the [demo code][ddos-detection] which will send server metrics in batches of 1000, and attempt to keep the timing roughly in line with the time deltas of the input records by sleeping between batches.
+
+The following steps will run the application:
+
+1. Set up Wallaroo for your OS following [these instructions][setup instructions]
+2. Check out the [Wallaroo Blog Exmaples GitHub repository][wallaroo blog examples] and navigate to the ddos-detection example directory
+
+    ```bash
+    git clone https://github.com/WallarooLabs/wallaroo_blog_examples.git
+    cd wallaroo_blog_examples/ddos-detection
+    ```
+
+3. Start the included listener in its own terminal so you can view the output
+
+    ```bash
+    python receiver.py --host 127.0.0.1:7002
+    ```
+
+4. Start two Wallaroo workers
+    in two separate terminals:
+    1. Initializer:
+
+        ```bash
+        export PYTHONPATH="$PYTHONPATH:.:$HOME/wallaroo-tutorial/wallaroo/machida"
+        export PATH="$PATH:$HOME/wallaroo-tutorial/wallaroo/machida/build"
+        machida --application-module ddos_detector --in 127.0.0.1:7010 --out 127.0.0.1:7002 \
+          --control 127.0.0.1:12500 --data 127.0.0.1:12501 --name initializer --cluster-initializer \
+          --worker-count 2 --metrics 127.0.0.1:5001 --ponythreads 1 --ponypinasio --ponynoblock
+        ```
+
+    2. 2nd Worker:
+
+        ```bash
+        export PYTHONPATH="$PYTHONPATH:.:$HOME/wallaroo-tutorial/wallaroo/machida"
+        export PATH="$PATH:$HOME/wallaroo-tutorial/wallaroo/machida/build"
+        machida --application-module ddos_detector --in 127.0.0.1:7010  --out 127.0.0.1:7003 \
+          --control 127.0.0.1:12500  --name worker --metrics 127.0.0.1:5001 --ponythreads 1 \
+          --ponypinasio --ponynoblock
+        ```
+
+4. Start the sender:
+
+    ```bash
+    python sender.py --host 127.0.0.1:7010 --file data.json --batch 1000
+    ```
+
+![demo gif][demo gif]
```

Comment:
love the gif but it looks weird on the page. i think there needs to be a line of intro text for it. like: 

"And you'll see output like:"